src/AggregateFunctions/Combinators/AggregateFunctionDistinct.h

```diff
@@ -20,59 +20,53 @@ struct AggregateFunctionDistinctSingleNumericData
     /// When creating, the hash table must be small.
     using Set = HashSetWithStackMemory<T, DefaultHash<T>, 4>;
     using Self = AggregateFunctionDistinctSingleNumericData<T>;
-    Set set;
 
-    void add(const IColumn ** columns, size_t /* columns_num */, size_t row_num, Arena *)
-    {
-        const auto & vec = assert_cast<const ColumnVector<T> &>(*columns[0]).getData();
-        set.insert(vec[row_num]);
-    }
+    /// history will hold all values added so far
+    Set history;
 
-    void merge(const Self & rhs, Arena *)
+    void add(const IColumn ** columns, size_t /* columns_num */, size_t row_num, MutableColumns & argument_columns, Arena *)
     {
-        set.merge(rhs.set);
-    }
+        const auto & vec = assert_cast<const ColumnVector<T> &>(*columns[0]).getData();
+        const T value = vec[row_num];
 
-    void serialize(WriteBuffer & buf) const
-    {
-        set.write(buf);
+        if (!history.contains(value))
```

Comment:
Here we will do a lookup twice. Can we use the return value of the insert method?

Reply:
Good catch. I've optimized this and the other similar spots as well.